pkg/runtime/wfengine/backend.go

```diff
@@ -0,0 +1,208 @@
+	// request. Note that this request goes directly to the actor runtime, bypassing the API layer.
+	req := invokev1.
+		NewInvokeMethodRequest(CreateWorkflowInstanceMethod).
+		WithActor(WorkflowActorType, workflowInstanceID).
+		WithRawData(eventData, invokev1.OctetStreamContentType)
+	if _, err := be.actors.Call(ctx, req); err != nil {
+		return err
+	}
+	return nil
+}
+
+// GetOrchestrationMetadata implements backend.Backend
+func (be *actorBackend) GetOrchestrationMetadata(ctx context.Context, id api.InstanceID) (*api.OrchestrationMetadata, error) {
+	// Invoke the corresponding actor, which internally stores its own workflow metadata
+	req := invokev1.
+		NewInvokeMethodRequest(GetWorkflowMetadataMethod).
+		WithActor(WorkflowActorType, string(id)).
+		WithRawData(nil, invokev1.OctetStreamContentType)
+	if res, err := be.actors.Call(ctx, req); err != nil {
+		return nil, err
+	} else {
+		_, data := res.RawData()
+		if len(data) == 0 {
+			return nil, api.ErrInstanceNotFound
+		}
+
+		// Decode the data using encoding/gob (https://go.dev/blog/gob)
+		buffer := bytes.NewBuffer(data)
+		dec := gob.NewDecoder(buffer)
+		var metadata api.OrchestrationMetadata
+		if err := dec.Decode(&metadata); err != nil {
+			return nil, err
+		}
+		return &metadata, nil
+	}
+}
+
+// AbandonActivityWorkItem implements backend.Backend
+func (*actorBackend) AbandonActivityWorkItem(context.Context, *backend.ActivityWorkItem) error {
+	panic("unimplemented")
+}
+
+// AbandonOrchestrationWorkItem implements backend.Backend
+func (*actorBackend) AbandonOrchestrationWorkItem(context.Context, *backend.OrchestrationWorkItem) error {
+	panic("unimplemented")
+}
+
+// AddNewOrchestrationEvent implements backend.Backend
+func (*actorBackend) AddNewOrchestrationEvent(context.Context, api.InstanceID, *backend.HistoryEvent) error {
+	panic("unimplemented")
+}
+
+// CompleteActivityWorkItem implements backend.Backend
+func (*actorBackend) CompleteActivityWorkItem(ctx context.Context, wi *backend.ActivityWorkItem) error {
+	// Resumes workflow execution code path in the actor
+	wi.Properties[CallbackChannelProperty].(chan bool) <- true
+	return nil
+}
+
+// CompleteOrchestrationWorkItem implements backend.Backend
+func (*actorBackend) CompleteOrchestrationWorkItem(ctx context.Context, wi *backend.OrchestrationWorkItem) error {
+	// Resumes workflow execution code path in the actor
+	wi.Properties[CallbackChannelProperty].(chan bool) <- true
+	return nil
+}
+
+// CreateTaskHub implements backend.Backend
+func (*actorBackend) CreateTaskHub(context.Context) error {
+	return nil
+}
+
+// DeleteTaskHub implements backend.Backend
+func (*actorBackend) DeleteTaskHub(context.Context) error {
+	panic("unimplemented")
+}
+
+// GetActivityWorkItem implements backend.Backend
+func (be *actorBackend) GetActivityWorkItem(ctx context.Context) (*backend.ActivityWorkItem, error) {
+	// Wait for the workflow actor to signal us with some work to do
+	select {
```

Comment:
Should we have a timeout in this select?

Reply:
That's fine as long as anything calling this expects that behavior :+1: - might it be worth documenting?